### PR TITLE
RSDK-5941 allow the SDK to request and receive video/h264 encoded frames

### PR DESCRIPTION
**Pull request by bazile-clyde**
~> This code will not compile until API proto changes are merged. See [TODO](#TODO) below.~

# Description 
These changes will request H264 frames from camera components that support them. These changes also create a new image.Image type, H264, that allows modules to pass H264 encoded bytes through the server downstream to the app for streaming.

# Testing
I tested this with the Go SDK and live stream using a custom module that could return both H264 and JPEG frames. The SDK can request "video/h264" or "image/JPEG" without issue. However, it still cannot decode "video/h264" so I could only ensure the server didn't panic. The live stream did work since the front end can decode H264 frames just fine. I also tested a regular old camera stream simultaneously to make sure this doesn't break anything. You can use my testing module, [camera-static](https://github.com/bazile-clyde/camera-static/tree/main), to test these changes as well. If you choose to, be sure to [create the test video](https://github.com/bazile-clyde/camera-static/tree/main) and update [the path](https://github.com/bazile-clyde/camera-static/blob/main/main.go#L130) to point towards it. Everything else should be straightforward. Ping me for questions.

[Screencast from 12-11-2023 05:17:40 PM.webm](https://github.com/viamrobotics/rdk/assets/34226620/cc60b80e-9ae0-41cb-94b2-30b12a30a7e7)

# TODO
- [x] Merge proto changes from https://github.com/viamrobotics/api/pull/408.

### CODE REVIEW

**Comment by bazile-clyde** [comment]
@seanavery the tests are all passing, and the race condition is fixed.

File: gostream/stream.go [hunk 1]
```diff
@@ -254,25 +256,33 @@ func (bs *basicStream) processInputFrames() {
...
 
-			bounds := framePair.Media.Bounds()
-			newDx, newDy := bounds.Dx(), bounds.Dy()
-			if bs.videoEncoder == nil || dx != newDx || dy != newDy {
-				dx, dy = newDx, newDy
-				bs.logger.Infow("detected new image bounds", "width", dx, "height", dy)
+			var encodedFrame []byte
+
+			if frame, ok := framePair.Media.(*rimage.LazyEncodedImage); ok && frame.MIMEType() == utils2.MimeTypeH264 {
+				encodedFrame = frame.RawData() // nothing to do; already encoded
```
Comment (seanavery) [inline gostream/stream.go:262]: nice :)

File: gostream/stream.go [hunk 2]
```diff
@@ -254,25 +256,33 @@ func (bs *basicStream) processInputFrames() {
...
+			} else {
+				bounds := framePair.Media.Bounds()
```
Comment (seanavery) [inline gostream/stream.go:264]: Maybe I am reading this wrong but I thought that `Bounds()` is not implemented for `H264` struct. What is `framepair.Media` here? Should we implement those getters?

File: rimage/image.go [hunk 3]
```diff
@@ -158,3 +158,36 @@ func ConvertColorImageToLuminanceFloat(img *Image) *mat.Dense {
...
+	GOPSize int
+}
+
+// ColorModel unimplemented.
+func (h H264) ColorModel() color.Model {
+	panic("not implemented")
+}
+
+// Bounds unimplemented.
+func (h H264) Bounds() image.Rectangle {
```
Comment (seanavery) [inline rimage/image.go:176]: Do we need this for updating width/height params for the stream?

**Review by seanavery — APPROVED** [review]
LGTM - straightforward and simple implementation and helpful pattern to signal which mimetypes are supported from a camera.

Only question is how the Bounds calls are working without the getter implemented.

**Pull request merged**